daffodil-core/src/main/scala/org/apache/daffodil/grammar/ModelGroupGrammarMixin.scala

```diff
@@ -42,7 +42,7 @@ trait ModelGroupGrammarMixin
   private lazy val groupRightFraming = prod("groupRightFraming") { TrailingSkipRegion(this) }
 
   final override lazy val termContentBody = prod("termContentBody") {
-    dfdlStatementEvaluations ~ groupLeftFraming ~ groupContentWithInitiatorTerminator ~ groupRightFraming
+    dfdlScopeBegin ~ dfdlStatementEvaluations ~ groupLeftFraming ~ groupContentWithInitiatorTerminator ~ groupRightFraming ~ dfdlScopeEnd
```

Comment:
Mabye this is the ordering you're talking about is incorrect? Do we need to split up dfdlStatementEvaluations into pattern and not pattern here? So the issue is on model groups and not elements?

Reply:
Yes, this is where I was looking at things. So I guess we do need a small fix here to do discriminators/asserts with test="pattern" first.

Reply:
Okay, so I've made a change to make this reflect what is described in the spec and process statements with patterns before newVariableInstance. I added a test that I think demonstrates this, but I am struggling to think of a case where an assert or discriminator statement could affect a newVariableInstance, especially if you can't use a variable to store the pattern used by the assert/discriminator. I tried to add a test doing that, but was unsuccessful.

Reply:
Hmm. An assert has a pattern and a message. Suppose the pattern fails. The message could be an expression that reads a variable defined by NVI with a default value. That would make it an error to subsequently set the variable.